File: sach/de3392.cpp
//DE3392.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
int m, n, s;   vector <vector <int>> a;
void sub1(vector <int> a, vector <int> b) {
    ll result = 0;
    sort(b.begin(),b.end());
    for(int x : a) {
        int u = lower_bound(b.begin(),b.end(),s-x)-b.begin();
        int v = upper_bound(b.begin(),b.end(),s-x)-b.begin();
        result += v - u;
    }cout << result;
}
void sub2() {
    vector<int> b,c;
    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            b.push_back(a[0][i] + a[1][j]);
    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            c.push_back(a[2][i] + a[3][j]);
    sub1(b,c);
}
void sub3() {
    vector<int> b,c;
    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            for(int k = 0; k < n; k++)
                b.push_back(a[2][i] + a[3][j] + a[4][k]);
    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            c.push_back(a[0][i] + a[1][j] );
    sub1(c,b);
}
int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    cin >> m >> n >> s;
    a.assign(m, vector <int> (n, 0));
    for(int i = 0; i < m; i++)
        for(auto &x: a[i]) cin >> x;
    if (m == 2) sub1(a[0],a[1]);
    else if (m == 4) sub2();
    else sub3();
}

File: sach/de3392.py
//DE3392.PY - Tran Huu Nam - hsgtin.vn - 2024
import bisect
from typing import List
m, n, s = 0, 0, 0
a: List[List[int]] = []
def sub1(a: List[int], b: List[int]) -> None:
    result = 0
    b.sort()
    for x in a:
        u = bisect.bisect_left(b, s - x)
        v = bisect.bisect_right(b, s - x)
        result += v - u
    print(result)
def sub2() -> None:
    b, c = [], []
    for i in range(n):
        for j in range(n):
            b.append(a[0][i] + a[1][j])
    for i in range(n):
        for j in range(n):
            c.append(a[2][i] + a[3][j])
    sub1(b, c)
def sub3() -> None:
    b, c = [], []
    for i in range(n):
        for j in range(n):
            for k in range(n):
                b.append(a[2][i] + a[3][j] + a[4][k])
    for i in range(n):
        for j in range(n):
            c.append(a[0][i] + a[1][j])
    sub1(c, b)
def main() -> None:
    global m, n, s, a
    m, n, s = map(int, input().split())
    a = [[0] * n for _ in range(m)]
    for i in range(m):
        a[i] = list(map(int, input().split()))
    if m == 2:
        sub1(a[0], a[1])
    elif m == 4:
        sub2()
    else:
        sub3()
if __name__ == "__main__":
    main()

